lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1417 @@
+      auto info = module.getDomainInfoAttr();
+      auto i = arg.getArgNumber();
+      return getDomainTypeID(info, i);
+    }
+
+    auto result = dyn_cast<OpResult>(value);
+    auto *owner = result.getOwner();
+    auto instance = cast<InstanceOp>(owner);
+    auto info = instance.getDomainInfoAttr();
+    auto i = result.getResultNumber();
+    return getDomainTypeID(info, i);
+  }
+
+private:
+  void processDomain(DomainOp op) {
+    auto index = domainTable.size();
+    auto name = op.getNameAttr();
+    domainTable.push_back(op);
+    typeIDTable.insert({name, index});
+  }
+
+  void processCircuit(CircuitOp circuit) {
+    for (auto decl : circuit.getOps<DomainOp>())
+      processDomain(decl);
+  }
+
+  /// A map from domain type ID to op.
+  SmallVector<DomainOp> domainTable;
+
+  /// A map from domain name to type ID.
+  DenseMap<StringAttr, DomainTypeID> typeIDTable;
+};
+
+/// Information about the changes made to the interface of a module, which can
+/// be replayed onto an instance.
+struct ModuleUpdateInfo {
+  /// The updated domain information for a module.
+  ArrayAttr portDomainInfo;
+  /// The domain ports which have been inserted into a module.
+  PortInsertions portInsertions;
+};
+
+struct GlobalState {
+  GlobalState(CircuitOp circuit) : circuitInfo(circuit) {}
+
+  CircuitDomainInfo circuitInfo;
+  DenseMap<StringAttr, ModuleUpdateInfo> moduleUpdateTable;
+};
+
+} // namespace
+
+//====--------------------------------------------------------------------------
+// Terms: Syntax for unifying domain and domain-rows.
+//====--------------------------------------------------------------------------
+
+namespace {
+
+/// The different sorts of terms in the unification engine.
+enum class TermKind {
+  Variable,
+  Value,
+  Row,
+};
+
+/// A term in the unification engine.
+struct Term {
+  constexpr Term(TermKind kind) : kind(kind) {}
+  TermKind kind;
+};
+
+/// Helper to define a term kind.
+template <TermKind K>
+struct TermBase : Term {
+  static bool classof(const Term *term) { return term->kind == K; }
+  TermBase() : Term(K) {}
+};
+
+/// An unknown value.
+struct VariableTerm : public TermBase<TermKind::Variable> {
+  VariableTerm() : leader(nullptr) {}
+  VariableTerm(Term *leader) : leader(leader) {}
+  Term *leader;
+};
+
+/// A concrete value defined in the IR.
+struct ValueTerm : public TermBase<TermKind::Value> {
+  ValueTerm(Value value) : value(value) {}
+  Value getValue() const { return value; }
+  Value value;
+};
+
+/// A row of domains.
+struct RowTerm : public TermBase<TermKind::Row> {
+  RowTerm(ArrayRef<Term *> elements) : elements(elements) {}
+  ArrayRef<Term *> elements;
+};
+
+/// A helper for assigning low numeric IDs to variables for user-facing output.
+struct VariableIDTable {
+  size_t get(VariableTerm *term) {
+    auto [it, inserted] = table.insert({term, table.size() + 1});
+    return it->second;
+  }
+
+  DenseMap<VariableTerm *, size_t> table;
+};
```

Comment:
Does the `table` need to be public? If not, then `class` with private table would be better.